pkg/zfs/plugin/backup.go

```diff
@@ -0,0 +1,251 @@
+			if err != nil {
+				p.Log.Errorf("zfs: Failed to get prev snapshot bkp %s err: {%v}", snapname, err)
+				return "", err
+			}
+		}
+	}
+
+	bkp, err := bkpbuilder.NewBuilder().
+		WithName(bkpname).
+		WithLabels(labels).
+		WithVolume(vol.Name).
+		WithPrevSnap(prevSnap).
+		WithSnap(snapname).
+		WithNode(vol.Spec.OwnerNodeID).
+		WithStatus(apis.BKPZFSStatusInit).
+		WithRemote(p.remoteAddr).
+		Build()
+
+	if err != nil {
+		return "", err
+	}
+	_, err = bkpbuilder.NewKubeclient().WithNamespace(p.namespace).Create(bkp)
+	if err != nil {
+		return "", err
+	}
+
+	return bkpname, err
+}
+
+func (p *Plugin) checkBackupStatus(bkpname string) {
+	bkpDone := false
+
+	for !bkpDone {
+		getOptions := metav1.GetOptions{}
+	        bkp, err := bkpbuilder.NewKubeclient().
+		        WithNamespace(p.namespace).Get(bkpname, getOptions)
+
+		if err != nil {
+			p.Log.Errorf("zfs: Failed to fetch backup info {%s}", bkpname)
+			p.cl.ExitServer = true
+			return
+		}
+
+		time.Sleep(backupStatusInterval * time.Second)
+
+		switch bkp.Status {
+		case apis.BKPZFSStatusDone, apis.BKPZFSStatusFailed, apis.BKPZFSStatusInvalid:
+			bkpDone = true
+			p.cl.ExitServer = true
+		}
+	}
+}
+
+func (p *Plugin) doBackup(volumeID string, snapname string, schdname string) (string, error) {
+
+	pv, err := p.getPV(volumeID)
+	if err != nil {
+		p.Log.Errorf("zfs: Failed to get pv %s snap %s schd %s err %v", volumeID, snapname, schdname, err)
+		return "", err
+	}
+
+	if pv.Spec.PersistentVolumeSource.CSI == nil {
+		return "", errors.New("zfs: err not a CSI pv")
+	}
+
+	volHandle := pv.Spec.PersistentVolumeSource.CSI.VolumeHandle
+
+        getOptions := metav1.GetOptions{}
+        vol, err := volbuilder.NewKubeclient().
+                WithNamespace(p.namespace).Get(volHandle, getOptions)
+	if err != nil {
+		return "", err
+	}
+
+	if pv.Spec.ClaimRef != nil {
+		// add source namespace in the label to filter it at restore time
+		if vol.Labels == nil {
+			vol.Labels = map[string]string{}
+		}
+		vol.Labels[VeleroNsKey] = pv.Spec.ClaimRef.Namespace
+	} else {
+		return "", errors.Errorf("zfs: err pv is not claimed")
+	}
+
+	// reset the exit server to false
+	p.cl.ExitServer = false
+
+	filename := p.cl.GenerateRemoteFilename(volumeID, snapname)
+	if filename == "" {
+		return "", errors.Errorf("zfs: error creating remote file name for backup")
+	}
+
+	err = p.uploadZFSVolume(vol, filename)
+	if err != nil {
+		return "", err
+	}
+
+	size, err := strconv.ParseInt(vol.Spec.Capacity, 10, 64)
+	if err != nil {
+		return "", errors.Errorf("zfs: error parsing the size %s", vol.Spec.Capacity)
+	}
+
+	// TODO(pawan) should wait for upload server to be up
+	bkpname, err := p.createBackup(vol, schdname, snapname)
+	if err != nil {
+		return "", err
+	}
+
+	go p.checkBackupStatus(bkpname)
+
+	p.Log.Infof("zfs: uploading Snapshot %s file %s", snapname, filename)
+	ok := p.cl.Upload(filename, size)
+	if !ok {
+		p.deleteBackup(bkpname)
```

Comment:
need to log error of deleteBackup for debugging purpose or manual cleanup.

Reply:
already loggin it inside the funtion.